token/client/src/token.rs

```diff
@@ -2212,13 +2204,308 @@ where
+    ) -> TokenResult<(T::Output, T::Output)> {
+        let account_info = if let Some(account_info) = account_info {
+            account_info
+        } else {
+            let account = self.get_account_info(source_account).await?;
+            let confidential_transfer_account =
+                account.get_extension::<ConfidentialTransferAccount>()?;
+            TransferAccountInfo::new(confidential_transfer_account)
+        };
+
+        let (
+            equality_proof_data,
+            ciphertext_validity_proof_data,
+            range_proof_data,
+            source_decrypt_handles,
+        ) = account_info
+            .generate_split_transfer_proof_data(
+                transfer_amount,
+                source_elgamal_keypair,
+                source_aes_key,
+                destination_elgamal_pubkey,
+                auditor_elgamal_pubkey,
+            )
+            .map_err(|_| TokenError::ProofGeneration)?;
+
+        let new_decryptable_available_balance = account_info
+            .new_decryptable_available_balance(transfer_amount, source_aes_key)
+            .map_err(|_| TokenError::AccountDecryption)?;
+
+        let transfer_instruction = confidential_transfer::instruction::transfer_with_split_proofs(
+            &self.program_id,
+            source_account,
+            destination_account,
+            &self.pubkey,
+            new_decryptable_available_balance.into(),
+            source_authority,
+            context_state_accounts,
+            &source_decrypt_handles,
+        )?;
+
+        let transfer_with_equality_and_ciphertext_validity = self
+            .confidential_transfer_create_equality_and_ciphertext_validity_proofs_for_transfer(
+                context_state_accounts,
+                &equality_proof_data,
+                &ciphertext_validity_proof_data,
+                Some(&transfer_instruction),
+                Some(source_authority_keypair),
+                equality_proof_account_keypair,
+                ciphertext_validity_proof_account_keypair,
+                context_state_authority_keypair,
+            );
+
+        let transfer_with_range_proof = self.confidential_transfer_create_range_proof_for_transfer(
+            context_state_accounts,
+            &range_proof_data,
+            Some(&transfer_instruction),
+            Some(source_authority_keypair),
+            range_proof_account_keypair,
+            context_state_authority_keypair,
+        );
+
+        try_join!(
+            transfer_with_equality_and_ciphertext_validity,
+            transfer_with_range_proof
+        )
+    }
+
+    /// Create equality and ciphertext validity proofs for a confidential transfer.
+    ///
+    /// If an optional transfer instruction is provided, then the transfer instruction is attached
+    /// to the same transaction.
+    #[allow(clippy::too_many_arguments)]
+    pub async fn confidential_transfer_create_equality_and_ciphertext_validity_proofs_for_transfer<
```

Comment:
nit: instead of `proofs` can you say `context_state` since that's what's being done? And rather than overloading this public function to optionally add a transfer, how about have three functions:

* `confidential_transfer_create_equality_and_ciphertext_validity_context_states_for_transfer` -- just the creation of the context state accounts
* `confidential_transfer_create_equality_and_ciphertext_validity_context_states_and_transfer_parallel` -- specifically for parallel execution, requires the transfer instruction
* some internal helper function that they both call into to create the proofs, up to you for naming!

Reply:
Yes, that is a great idea!